### PR TITLE
Rust 1.51.0

### PR DESCRIPTION
**Pull request by benma**

### CODE REVIEW

**Comment by benma** [comment]
The firmware decreases in size a little with the new toolchain, but the bootloader increases significantly (55760 vs 47728 bytes).


[bootloaders.tar.gz](https://github.com/digitalbitbox/bitbox02-firmware/files/6263282/bootloaders.tar.gz)

Here are the .bin and .elf files for both.

**Comment by benma** [comment]
@x1ddos added a commit to workaround the issue until it is fixed in the compiler.

**Comment by x1ddos** [comment]
:+1: pulling the new container image...

File: src/rust/bitbox02-rust/src/bb02_async.rs [hunk 1]
```diff
@@ -48,8 +48,8 @@ impl<O> core::future::Future for AsyncOption<'_, O> {
 
 /// Waits for an option to contain a value and returns that value, leaving `None` in its place.
 /// E.g. `assert_eq!(option(&Some(42)).await, 42)`.
-pub fn option<'a, O>(option: &'a RefCell<Option<O>>) -> AsyncOption<'a, O> {
-    AsyncOption(&option)
+pub fn option<O>(option: &RefCell<Option<O>>) -> AsyncOption<'_, O> {
```
Comment (x1ddos) [inline src/rust/bitbox02-rust/src/bb02_async.rs:51]: What's the reason for `'_` in returned value? I tried without it. rust seems to be able to figure it out without issues and clippy doesn't complain.

File: src/rust/bitbox02-rust/src/bb02_async.rs [hunk 2]
```diff
@@ -48,8 +48,8 @@ impl<O> core::future::Future for AsyncOption<'_, O> {
...
+    AsyncOption(option)
```
Comment (x1ddos) [inline src/rust/bitbox02-rust/src/bb02_async.rs:52]: just checking: this was not clippy, right.

**Review by x1ddos — COMMENTED** [review]
Can we compare hashes, just to make sure it's still byte-to-byte reproducible. Mine are:

```
$ git rev-parse HEAD
bf832fc89baabe780478c8bfbfb5293cd53d875e
$ git status
On branch rust-1.51.0
nothing to commit, working tree clean

$ make -j firmware; make -j firmware-btc; make -j bootloader
=============================================
--             - General -
-- Firmware version:       v9.5.0-pre+bf832fc89b
-- Firmware v. (git):      v9.5.0+6
-- Firmware-btc version:   v9.5.0-pre+bf832fc89b
-- Firmware-btc v. (git):  v9.5.0+6
-- Bootloader version:     v1.0.3-pre+bf832fc89b
-- Bootloader v. (git):    v1.0.3+250
...

$ sha256sum build/bin/*.bin
04249f530c39260aef809147fc7cd8c3c60f0be810cababb7aadf360047488c5  build/bin/bootloader.bin
a1f741e6e6dd31f88b664b8d1f4276aca13e22651e2ac953ea0a403697259579  build/bin/firmware-btc.bin
e6527c62105a028b8384079fca8635b76aa5c0369f7389681035fd426245b2b8  build/bin/firmware.bin
```

File: src/rust/bitbox02-rust/src/bb02_async.rs (hunk 2 above)
Comment (benma) [reply src/rust/bitbox02-rust/src/bb02_async.rs:52]: I guess not. I really don't like that double reference somehow just works. The previous version should have been a compilation error imho.

File: src/rust/util/src/ascii.rs [hunk 3]
```diff
@@ -29,7 +29,7 @@ pub fn is_printable_ascii<T: AsRef<[u8]>>(bytes: T, charset: Charset) -> bool {
     bytes
         .as_ref()
         .iter()
-        .all(|&b| (b >= 32 && b <= 126) || (charset == Charset::AllNewline && b == b'\n'))
+        .all(|&b| (32..=126).contains(&b) || (charset == Charset::AllNewline && b == b'\n'))
```
[the remark this answers is not in the document]
Comment (benma) [reply src/rust/util/src/ascii.rs]: I find it quite readable, but I reverted it and disabled that linter.

File: src/rust/bitbox02-rust/src/bb02_async.rs (hunk 1 above)
Comment (benma) [reply src/rust/bitbox02-rust/src/bb02_async.rs]: I might have copied it from the original clippy fix suggestion?

I am surprised it works without it, after all, it is an explicit param in the declaration of AsyncOption...

Anyway, removed.

**Review by x1ddos — APPROVED** [review]
LGTM

File: src/rust/bitbox02-rust/src/bb02_async.rs (hunk 2 above)
Comment (NickeZ) [reply src/rust/bitbox02-rust/src/bb02_async.rs:52]: you are not alone, however auto-deref makes the language much more ergonomic: https://stackoverflow.com/questions/28519997/what-are-rusts-exact-auto-dereferencing-rules

**Pull request merged**